btf/btf.go

```diff
@@ -698,24 +698,24 @@ func LoadSplitSpecFromReader(r io.ReaderAt, base *Spec) (*Spec, error) {
 
 // TypesIterator iterates over types of a given spec.
 type TypesIterator struct {
-	spec  *Spec
+	types []Type
 	index int
 	// The last visited type in the spec.
 	Type Type
 }
 
 // Iterate returns the types iterator.
```

Comment:
I guess you mean document them somewhere, like Spec.Add? That would make sense, not sure what you mean by `changes to ... will poison the Spec` though. OK to discuss in a separate PR?

Reply:
Right now, calling `spec.Iterate()` and modifying any `Type` returned by `spec.TypeByXYZ` will have the changes seen by the iterator, until we make all btf.Spec queries return copies of `Type`s.

Reply:
Ah, yes! In my mind the behaviour should be:

- Types are mutable via Iterate(), etc.
- Specs are mutable via Add()
- Iterate() will only visit types that have been Add()ed before

Maybe this is too subtle? We could drop the last point, but that seems like we're leaking an implementation detail of Add.

> until we make all btf.Spec queries return copies of `Type`s.

Are you referring to the lazy copy we talked about, or something else?

Reply:
> Types are mutable via Iterate(), etc.

Yes, this should actually be a feature to allow fixups across a whole Spec if needed.

> Are you referring to the lazy copy we talked about

Yep! But given the above point, should `TypeBy*()` methods really copy? e.g. what if I want to modify a particular Type obtained using a query and commit the changes to the Spec? Spec.Add() isn't of much use to the caller for modifying existing types.

For example, under lazy copying, I wouldn't be able to do `s.TypeByName(".kconfig", &ds)` and fudge `ds`. I'd basically have to use an iterator and check on type/name, basically doing the heavy lifting myself. The difference between a query (`TypeBy*`) and the iterator might be too subtle, but maybe that's okay.

---

(devil's advocate)

The point I was trying to make during our call wrt. being 'too defensive': It might be a losing battle (in the long term) trying to make sure all `Type`s we return are actually copies. Due to the expensive nature of copies, they may best be left as an explicit operation the caller needs to make. By default, things mutate, even the (now shared) vmlinux spec.

I would broaden the first point in your list of principles:

- Types obtained[^1] from a btf.Spec are mutable, any changes will take effect when loaded

Also, all types implementing `Type` are pointer types, so it's rather self-explanatory that changing them might affect things elsewhere. If one package is being a bad citizen by mutating types in the shared vmlinux spec, the fix is straightforward: include a `Spec.Copy()` or `btf.Copy(Type)`. If this property is well-documented, it shouldn't cause too many surprises.

[^1]: through `Iterate()`, `TypeBy*()` and any future APIs

Reply:
> I wouldn't be able to do s.TypeByName(".kconfig", &ds) and fudge ds.

You would :) Lazy copy shouldn't change semantics, let's discuss this when I create the PR.

> I would broaden the first point in your list of principles:

Agreed.

> If one package is being a bad citizen by mutating types in the shared vmlinux spec, the fix is straightforward: include a Spec.Copy() or btf.Copy(Type). If this property is well-documented, it shouldn't cause too many surprises.

I disagree here, it's fine if the buggy package breaks, it's not OK if other packages are affected. Again, probably best discussed on the lazy copy PR.